### PR TITLE
CPPAPI:  Add support for Session API hooks.

### PR DESCRIPTION
**Pull request by ywkaras**
This is a very preliminary work in progress.  It does not currently compile.  Please only look at the changes to header files at this point.  Feedback would be very helpful.

### CODE REVIEW

File: include/tscpp/api/GlobalPlugin.h [hunk 1]
```diff
@@ -82,7 +83,15 @@ class GlobalPlugin : public Plugin
...
-  GlobalPluginState *state_; /**< Internal state tied to a GlobalPlugin */
+  struct PluginState;
+  PluginState *state_; /**< Internal state tied to a GlobalPlugin */
 };
 
+bool RegisterGlobalPlugin(const char *name, const char *vendor, const char *email);
+inline bool
+RegisterGlobalPlugin(std::string const &name, std::string const &vendor, std::string const &email)
+{
+  return RegisterGlobalPlugin(name.c_str(), vendor.c_str(), email.c_str());
```
Comment (SolidWallOfCode) [inline include/tscpp/api/GlobalPlugin.h]: Shouldn't these be either `std::string const&` or `std::string_view`? Why bother with `c_str()`?

File: include/tscpp/api/GlobalPluginHooks.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ * @see TransactionPlugin
+ */
+class GlobalPluginHooks : public SessionPluginHooks
+{
+public:
+  /**
+   * A enumeration of the available types of Hooks. These are used with GlobalPlugin::registerHook().
+   */
+  enum HookType {
+    HOOK_SSN_START = 0, /**< This hook will be fired after Session start. */
```
Comment (SolidWallOfCode) [inline include/tscpp/api/GlobalPluginHooks.h]: `HOOK_SSN_CLOSE`?

File: include/tscpp/api/SessionPluginHooks.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+ */
+class SessionPluginHooks : public TransactionPluginHooks
+{
+public:
+  /**
+   * A enumeration of the available types of Hooks. These are used with GlobalPlugin::registerHook()
+   * and SessionPlugin::registerHook().
+   */
+  enum HookType {
+    HOOK_TXN_START = 0, /**< This hook will be fired after transaction has started. */
```
Comment (SolidWallOfCode) [inline include/tscpp/api/SessionPluginHooks.h]: `TXN`?

**Comment by SolidWallOfCode** [comment]
I'm puzzled about the transaction start hooks in `SessionPluginHook`. Does setting a callback on the transaction start hook require a session object?

File: include/tscpp/api/GlobalPluginHooks.h (hunk 2 above)
Comment (ywkaras) [reply include/tscpp/api/GlobalPluginHooks.h]: In the CPPAPI you don't explicitly hooks transaction or session closes.  These hooks are used internally, you put close actions in the destructor of your plugin derived type.

**Comment by ywkaras** [comment]
[approve ci Clang-Analyzer]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
The first cut of the coding for this is done, next step is additional Au tests.

File: include/tscpp/api/GlobalPlugin.h (hunk 1 above)
Comment (ywkaras) [reply include/tscpp/api/GlobalPlugin.h]: You mean eliminate the first overload of RegisterGlobalPlugin()?

File: include/tscpp/api/SessionPluginHooks.h (hunk 3 above)
Comment (ywkaras) [reply include/tscpp/api/SessionPluginHooks.h]: Yup  https://github.com/apache/trafficserver/blob/master/include/ts/apidefs.h.in#L227

**Comment by ywkaras** [comment]
ALERT:  Changes are anticipated to the TS API as server-side session support is changed.  This PR may conflict with those changes.  Minimally, force a re-run of CI before merging this to master.

**Comment by ywkaras** [comment]
@calavera  This is a CPPAPI PR that's been pending for a long time.

**Comment by ywkaras** [comment]
I am closing this, since it's been open for almost a year.  Presumably, CPPAPI users have no need for session hooks, or are happy using them via the C plugin API.

**Pull request closed without merging**